private_gpt/server/ingest/ingest_service.py

```diff
@@ -81,84 +143,72 @@ def __init__(
+            # Load the index with store_nodes_override=True to be able to delete them
             index = load_index_from_storage(
                 storage_context=self.storage_context,
                 service_context=self.ingest_service_context,
                 store_nodes_override=True,  # Force store nodes in index and document stores
                 show_progress=True,
             )
-            for doc in documents:
-                index.insert(doc)
         except ValueError:
-            # Or create a new one if there is none
-            VectorStoreIndex.from_documents(
-                documents,
+            # There are no index in the storage context, creating a new one
+            logger.info("Creating a new vector store index")
+            index = VectorStoreIndex.from_documents(
+                [],
                 storage_context=self.storage_context,
                 service_context=self.ingest_service_context,
                 store_nodes_override=True,  # Force store nodes in index and document stores
                 show_progress=True,
             )
+            index.storage_context.persist(persist_dir=local_data_path)
+        return index
+
+    def _save_index(self) -> None:
+        self._index.storage_context.persist(persist_dir=local_data_path)
+
+    def ingest(self, file_name: str, file_data: AnyStr | Path) -> list[IngestedDoc]:
+        logger.info("Ingesting file_name=%s", file_name)
+        documents = _transform_file_into_documents(file_name, file_data)
+        logger.info(
+            "Transformed file=%s into count=%s documents", file_name, len(documents)
+        )
+        logger.debug("Saving the documents in the index and doc store")
+        return self._save_docs(documents)
+
+    def bulk_ingest(self, files: list[tuple[str, Path]]) -> list[IngestedDoc]:
+        logger.info("Ingesting file_names=%s", [f[0] for f in files])
 
+        with multiprocessing.Pool(processes=BULK_INGEST_WORKER_NUM) as pool:
+            documents = list(
+                itertools.chain.from_iterable(
+                    pool.starmap(_transform_file_into_documents, files)
+                )
+            )
+        logger.info(
+            "Transformed count=%s files into count=%s documents",
+            len(files),
+            len(documents),
+        )
+        return self._save_docs(documents)
+
+    def _save_docs(self, documents: list[Document]) -> list[IngestedDoc]:
+        for document in documents:
+            document.metadata["doc_id"] = document.doc_id
+            # We don't want the Embeddings search to receive this metadata
+            document.excluded_embed_metadata_keys = ["doc_id"]
+            # We don't want the LLM to receive these metadata in the context
+            document.excluded_llm_metadata_keys = ["file_name", "doc_id", "page_label"]
+        logger.debug("Inserting the documents in the index")
+        for doc in documents:
+            self._index.insert(doc)
```

Comment:
This is the new bottleneck now. This is CPU bound -- currently running on a single core